### PR TITLE
`HtmlImage.toByteData`: handle blob rawUnmodified format / add png support

### PR DESCRIPTION
**Pull request by maRci002**
Resolves https://github.com/flutter/flutter/issues/119065.

This PR adds the following features:

- If `DomHTMLImageElement.src` is `blob` and `format = ui.ImageByteFormat.rawUnmodified` then we can restore original bytes
- `ui.ImageByteFormat.png` format is handled properly

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/html_image_codec.dart [hunk 1]
```diff
@@ -194,16 +194,54 @@ class HtmlImage implements ui.Image {
...
+        }
+
+        final DomCanvasElement canvas = createDomCanvasElement()
+          ..width = width.toDouble()
+          ..height = height.toDouble();
+        final DomCanvasRenderingContext2D ctx = canvas.context2D;
+        ctx.drawImage(imgElement, 0, 0);
+        // canvas.toDataURL defaults to: 'image/png'
+        final String dataUrl = canvas.toDataURL();
+        return _byteDataFromDataString(dataUrl);
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/html_image_codec.dart:209]: This might be useful for CanvasKit as well. cc @yjbanov

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/html_image_codec.dart:213]: Yep, but I'm OK taking one step at a time.

File: lib/web_ui/lib/src/engine/html_image_codec.dart [hunk 2]
```diff
@@ -194,16 +198,44 @@ class HtmlImage implements ui.Image {
...
+          return _byteDataFromDataString(imgElement.src!);
+        }
+
+        final DomCanvasElement canvas = createDomCanvasElement()
+          ..width = width.toDouble()
+          ..height = height.toDouble();
+        final DomCanvasRenderingContext2D ctx = canvas.context2D;
+        ctx.drawImage(imgElement, 0, 0);
+        // canvas.toDataURL defaults to: 'image/png'
+        final String dataUrl = canvas.toDataURL();
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html_image_codec.dart:212]: After reading the image data back, set `canvas.width` and `canvas.height` to zero. Safari on iOS will crash if the GC doesn't collect 2D canvases fast enough. Setting the image size to zero causes Safari to free internal image memory immediately.

**Review by yjbanov — CHANGES_REQUESTED** [review]

**Comment by Hixie** [comment]
@maRci002 Will you be able to continue working on this? It's totally fine if not; we can take it over (though it might take longer). I'm just asking because I'm trying to clean out our review queue.

**Comment by maRci002** [comment]
> @maRci002 Will you be able to continue working on this? It's totally fine if not; we can take it over (though it might take longer). I'm just asking because I'm trying to clean out our review queue.

Unfortunately I don't have any free time in the near future.

I think this part of the PR is invalid, since it should return raw decoded bytes and not the encoded ones.

> * If `DomHTMLImageElement.src` is `blob` and `format = ui.ImageByteFormat.rawUnmodified` then we can restore original bytes

**Pull request closed without merging**